### PR TITLE
core/qbft: refactor to explicit justifications

### PR DESCRIPTION
**Pull request by corverroos**
Refactors QBFT to explicit justifications.

Also:
 - Do not exit when decided, keep running until context cancelled
 - Send Qcommit to others after decided
 - Jump ahead when receiving justified messages
 - Trim old round messages from buffer
 - Add constant period tests
 - Make instance type bytes (to support "any" instance type)

category: refactor 
ticket: #445 
feature_set: alpha

### CODE REVIEW

**Review by dB2510 — APPROVED** [review]

**Review by xenowits — APPROVED** [review]

File: core/qbft/qbft.go [hunk 1]
```diff
@@ -87,18 +99,21 @@ type Msg struct {
...
+	uponQuorumCommits
+	uponUnjustRoundChange
+	uponFPlus1RoundChanges
+	uponQuorumRoundChanges
 )
 
-// Run returns the consensus decided value (Qcommit) or a context closed error.
-func Run(ctx context.Context, d Definition, t Transport, instance, process int64, inputValue []byte) (value []byte, err error) {
+// Run executes the consensus algorithm until the context closed.
+//nolint:gocognit // If is indeed a complex algorithm.
```
Comment (dB2510) [inline core/qbft/qbft.go:113]:
```suggestion
//nolint:gocognit // It is indeed a complex algorithm.
```

File: core/qbft/qbft.go [hunk 2]
```diff
@@ -109,167 +124,264 @@ func Run(ctx context.Context, d Definition, t Transport, instance, process int64
...
+	// === State ===
+
+	var (
+		round           int64 = 1
+		preparedRound   int64
+		preparedValue   []byte
+		preparedJustify []Msg
+		qCommit         []Msg
+		buffer          []Msg
+		dedupIn         = make(map[dedupKey]bool)
```
Comment (dB2510) [inline core/qbft/qbft.go:136]: What is dedup?

File: core/qbft/qbft.go [hunk 3]
```diff
@@ -87,18 +99,21 @@ type Msg struct {
...
+	uponUnjustRoundChange
+	uponFPlus1RoundChanges
+	uponQuorumRoundChanges
 )
 
-// Run returns the consensus decided value (Qcommit) or a context closed error.
-func Run(ctx context.Context, d Definition, t Transport, instance, process int64, inputValue []byte) (value []byte, err error) {
+// Run executes the consensus algorithm until the context closed.
+//nolint:gocognit // If is indeed a complex algorithm.
+func Run(ctx context.Context, d Definition, t Transport, instance []byte, process int64, inputValue []byte) (err error) {
```
Comment (dB2510) [inline core/qbft/qbft.go:114]: Suggest: parameter names can be `def Defination, trn Transport`

File: core/qbft/qbft.go [hunk 4]
```diff
@@ -72,13 +75,22 @@ const (
 
 // Msg defines the inter process messages.
```
Comment (dB2510) [inline core/qbft/qbft.go:73]: suggest: replace word `process` with `nodes` as it is confusing since we deal with Charon **nodes** not processes

Comment (corverroos) [reply core/qbft/qbft.go:73]: the algorithm paper talks about processes (so sticking with its terminology)

File: core/qbft/qbft.go (hunk 3 above)
Comment (corverroos) [reply core/qbft/qbft.go:114]: sure, but `d` and `t` also works no?

File: core/qbft/qbft.go (hunk 2 above)
Comment (corverroos) [reply core/qbft/qbft.go:136]: A very common term for "deduplication"

File: core/qbft/qbft.go (hunk 1 above)
Comment (corverroos) [reply core/qbft/qbft.go:113]: will fix in generics PR 👍

File: core/qbft/qbft.go (hunk 3 above)
Comment (dB2510) [reply core/qbft/qbft.go:114]: def sounds more of defination and t confuses with t of testing

File: core/qbft/qbft.go (hunk 4 above)
Comment (dB2510) [reply core/qbft/qbft.go:73]: yeah sure, paper talks about processes but we are dealing with nodes not processes so we can/should modify terminologies as per our usecases

**Pull request merged**